pkg/local/sync.go

```diff
@@ -300,20 +312,40 @@ func (s *SyncManager) upload(ctx context.Context, rootPath string, remote *uri.U
 	remotePath := strings.TrimRight(remote.GetPath(), uri.PathSeparator)
 	dest := strings.TrimPrefix(filepath.ToSlash(fmt.Sprintf("%s%s%s", remotePath, uri.PathSeparator, path)), uri.PathSeparator)
 
-	f, err := os.Open(source)
+	fileStat, err := os.Lstat(source)
 	if err != nil {
 		return err
 	}
-	defer func() {
-		_ = f.Close()
-	}()
 
-	fileStat, err := f.Stat()
-	if err != nil {
-		return err
+	var contentSize int64
+	mtimeString := strconv.FormatInt(fileStat.ModTime().Unix(), 10)
+	metadata := map[string]string{
+		apiutil.ClientMtimeMetadataKey: mtimeString,
 	}
+	var contentReader io.ReadSeeker
 
-	b := s.progressBar.AddReader(fmt.Sprintf("upload %s", path), fileStat.Size())
+	handleSymlink := s.cfg.SymlinkSupport && fileStat.Mode()&os.ModeSymlink != 0
+	if handleSymlink {
+		// If symlink support is enabled, upload as symlink
+		target, err := os.Readlink(source)
+		if err != nil {
+			return fmt.Errorf("failed to read symlink target: %w", err)
+		}
+		metadata[apiutil.SymlinkMetadataKey] = target
+		contentSize = 0
+		contentReader = bytes.NewReader([]byte{})
+	} else {
```

Comment:
TOCTTOU issue here: if I manage to introduce a _between_ the stat call on l. 315 and the open call on l. 339, I can make you upload any file.

For instance, you lakectl local a directory, and mount just that directory inside a Docker container on /local.  The container cannot see /etc/ on the host.  Now you run my program inside that container.  I race against your sync to insert a `ln -s /etc/passwd /local/README.md` _between_ the stat and the open.  **Usually** I lose the race; but when I win it, you will upload `/etc/passwd` to the lakeFS repo.

You can probably fix this one by first opening the file, and then using fstatat on it with no-follow.

Reply:
This attack is not related to the current feature. We do not assume that the file system is locked when we perform lakectl local.
While we sync we perform diff between the filesystem and the remote and send command to upload/download.
So switching a regular to symlink and cause lakectl to follow it exists, without the support for symlinks.
Not saying we can't do something about it - but I think it is out of scope.